mitiq/cdr/clifford_training_data.py

```diff
@@ -0,0 +1,516 @@
+    Returns:
+        List[circ.Circuits]: list of near-Clifford circuits constructed from
+                             the circuits of interest.
+        List[List[float]]: list of list of angles that were replaced in each
+                           training circuit.
+        List[List[float]]: list of list of angles that were inserted in each
+                           training circuit.
+    """
+    circuits_list = []
+    # find all the non-Clifford gates:
+    # all_cliff = np.column_stack((not_rz_circ_data, rz_cliff))
+    angles_original_list = []
+    angles_replaced_list = []
+    # setting the seed:
+    if isinstance(random_state, int):
+        random_state = np.random.RandomState(random_state)
+    else:
+        random_state = randint(10 ** (8))
+        random_state = np.random.RandomState(random_state)
+    # generating a list of seeds used for each training circuit construction:
+    random_states = random_state.randint(
+        10000 * num_training_circuits, size=num_training_circuits
+    )
+    for n in range(num_training_circuits):
+        random_state = random_states[n]
+        random_state = np.random.RandomState(random_state)
+        # Convert data arry into cirq circuit and append it to the storage
+        #  array:
+        if additional_options:
+            (
+                projected_circuit,
+                angles_original,
+                angles_replaced,
+            ) = _map_to_near_clifford(
+                circuit,
+                fraction_non_clifford,
+                random_state,
+                method_select,
+                method_replace,
+                additional_options=additional_options.get(
+                    "additional_options"
+                ),
+            )
+        else:
+            (
+                projected_circuit,
+                angles_original,
+                angles_replaced,
+            ) = _map_to_near_clifford(
+                circuit,
+                fraction_non_clifford,
+                random_state,
+                method_select,
+                method_replace,
+            )
+
+        circuits_list.append(projected_circuit)
+        # this information is to make sure the probabilistic methods are
+        #  working as expected:
+        angles_original_list.append(angles_original)
+        angles_replaced_list.append(angles_replaced)
+    return circuits_list, angles_original_list, angles_replaced_list
+
+
+def _map_to_near_clifford(
+    circuit: Circuit,
+    fraction_non_clifford: float,
+    random_state: np.random.RandomState,
+    method_select: str = "random",
+    method_replace: str = "closest",
+    **additional_options: dict,
+) -> Tuple[Circuit, List[float], List[float]]:
+    """Function to take the information in some circuit of interest and
+        return a near-Clifford circuit as constructed according to
+        some user defined methods.
+    Args:
+        circuit: cirq.Circuit object of origional circuit of interest.
```

Comment:
Can just say "Circuit of interest" or "Circuit to map to a near-clifford circuit." No need to specify it is a cirq.Circuit because this is in the type annotation.

Should list all assumptions about the circuit here, though. Such as: we assume it only has single-qubit + CNOT gates, right?